crates/hir_ty/src/ctx.rs

```diff
@@ -344,7 +346,7 @@ impl InferenceCtx<'_> {
                             )
                         } else {
                             ExprMutability::ImmutableRef(
-                                current_bodies!(self).range_for_expr(local_def.value),
+                                current_bodies!(self).get_range_for_expr(local_def.value),
```

Comment:
I think that if the value of the local def doesn't exist, we should probably just lie and say it's mutable. An error of "you can't mutate this because it's an immutable reference" doesn't make sense because that's not the real underlying error.

Besides, a value of `ImmutableRef(None)` seems to have the same effect as `Mutable`. On line 78 they both do the same thing.

I'd probably change this whole match block to something like
```rust
match local_ty.as_pointer() {
    // Checking for bindings first makes more sense than what I wrote before
    // What do you think? A few tests would probably have to change.
    // Not really the focus of this PR though so it's wtv
    _ if !local_def.mutable => ExprMutability::ImmutableBinding(local_def.range),
    Some((mutable, _)) if assignment => {
        if mutable {
            ExprMutability::NotMutatingRefThroughDeref(
                current_bodies!(self).range_for_expr(expr),
            )
        } else if current_bodies!(self)[local_def.value] != Expr::Missing {
            ExprMutability::ImmutableRef(current_bodies!(self).range_for_expr(expr))
        } else {
            ExprMutability::Mutable
        }
    }
    _ => ExprMutability::Mutable,
}
```

Or at least something along these lines :)

Reply:
yeah, thank you

Reply:
the failing tests don't seem to indicate worse diagnistics, so I will update them.

the `assign_to_immutable_ref_no_deref` test would previously output a 
`TyDiagnosticHelpKind::ImmutableRef` and now outputs a `TyDiagnosticHelpKind::ImmutableBinding`

```capy
                foo :: () {
                    bar :: 5;

                    baz :: ^bar;

                    baz = 25;
                }
```
which imo is more intuitive. 

same for the the ` assign_to_mut_ref_no_deref` test 
```capy
                foo :: () {
                    bar := 5;

                    baz :: ^mut bar;

                    baz = 25;
                }
```
would previously produce `TyDiagnosticHelpKind::NotMutatingRefThroughDeref` and now produces `TyDiagnosticHelpKind::ImmutableBinding`.